### PR TITLE
Add support for structural configs

### LINKED ISSUE #207: Add structured configuration
We are missing [Structured Configuration](https://www.pulumi.com/docs/concepts/config/#structured-configuration) in the API

Add:
- `def requireJson(key: NonEmptyString): Output[JsValue]`
- `def requireObject[A: JsonFormat](key: (non-empty?)String): Output[A] = requireJson(key).map(_.convertTo[A])`

Relates to #204

### PR DESCRIPTION
**Pull request by pawelprazak**
- add support for JSON configs, with AST and object results
- code and docs cleanup

Fixes #207

### CODE REVIEW

**Comment by lbialy** [comment]
beside questions and nitpick about exports this looks awesome!

File: core/src/test/scala/besom/internal/ConfigTest.scala [hunk 1]
```diff
@@ -1,70 +1,213 @@
 package besom.internal
 
-import besom.*
+import besom._
```
[the remark this answers is not in the document]
Comment (pawelprazak) [reply core/src/test/scala/besom/internal/ConfigTest.scala:3]: if this is the case we need a tool that will enforce that

[the remark this answers is not in the document]
Comment (pawelprazak) [reply core/src/test/scala/besom/internal/ConfigTest.scala:3]:
```suggestion
import besom.*
```

File: core/src/main/scala/besom/internal/ConfigValueReader.scala [hunk 2]
```diff
@@ -1,5 +1,7 @@
 package besom.internal
 
+import spray.json._
```
Comment (pawelprazak) [inline core/src/main/scala/besom/internal/ConfigValueReader.scala:3]:
```suggestion
import spray.json.*
```

File: core/src/test/scala/besom/internal/ConfigTest.scala [hunk 3]
```diff
@@ -1,70 +1,213 @@
 package besom.internal
 
 import besom.*
+import spray.json._
```
Comment (pawelprazak) [inline core/src/test/scala/besom/internal/ConfigTest.scala:4]:
```suggestion
import spray.json.*
```

File: core/src/test/scala/besom/internal/ConfigTest.scala [hunk 4]
```diff
@@ -1,70 +1,213 @@
 package besom.internal
 
 import besom.*
+import spray.json.*
+import spray.json.DefaultJsonProtocol._
```
Comment (pawelprazak) [inline core/src/test/scala/besom/internal/ConfigTest.scala:5]:
```suggestion
import spray.json.DefaultJsonProtocol.*
```

File: core/src/test/scala/besom/internal/ConfigTest.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (lbialy) [reply core/src/test/scala/besom/internal/ConfigTest.scala]: https://scalameta.org/scalafmt/docs/configuration.html#rewritescala3converttonewsyntax

**Review by lbialy — APPROVED** [review]

**Pull request merged**